certgen/certgen.go

```diff
@@ -0,0 +1,164 @@
+package certgen
+
+import (
+	"crypto"
+	crand "crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509/pkix"
+	"github.com/google/monologue/interval"
+)
+
+var timeNowUTC = func() time.Time {
+	return time.Now().UTC()
+}
+
+// CertificateConfig contains details to be used to populate newly created leaf
+// certificates.
+type CertificateConfig struct {
+	// Required fields
+	//
+	// What these are set to, including the zero values if left unset, is what
+	// will appear in the leaf certificates.
+	SubjectCommonName         string
+	SubjectOrganization       string
+	SubjectOrganizationalUnit string
+	SubjectLocality           string
+	SubjectCountry            string
+	SignatureAlgorithm        x509.SignatureAlgorithm
+
+	// Optional fields
+
+	// Prefix is a prefix that will be used in conjunction with the
+	// SubjectCommonName to create a more specific DNS SAN.
+	Prefix string
+	// NotAfterInterval specifies an interval in which the NotAfter time of a
+	// certificate must fall.
+	NotAfterInterval *interval.Interval
+}
+
+// CA is a Certificate Authority that issues synthetic certificates and
+// certificate chains using its RootCert and RootKey.
+type CA struct {
+	RootCert   *x509.Certificate
+	RootKey    crypto.Signer
+	CertConfig CertificateConfig
+}
+
+// IssueCertificate creates a new leaf certificate, issued by the root specified
+// in the RootCert and RootKey fields of the CA, and configured using the
+// CertConfig in the CA.
+func (ca *CA) IssueCertificate() (*x509.Certificate, error) {
+	key, err := rsa.GenerateKey(crand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("error generating key pair: %s", err)
+	}
+
+	template, err := leafTemplate(ca.CertConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf template: %s", err)
+	}
+
+	leafDER, err := x509.CreateCertificate(crand.Reader, template, ca.RootCert, key.Public(), ca.RootKey)
+	if err != nil {
+		return nil, fmt.Errorf("error creating leaf certificate: %s", err)
+	}
+
+	leaf, err := x509.ParseCertificate(leafDER)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing leaf certificate DER: %s", err)
+	}
+
+	return leaf, nil
+}
+
+func leafTemplate(c CertificateConfig) (*x509.Certificate, error) {
+	sn, err := randSerialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	notAfter := randNotAfter(c.NotAfterInterval)
+
+	return &x509.Certificate{
+		SerialNumber: sn,
+		Subject: pkix.Name{
+			Country:            []string{c.SubjectCountry},
+			Organization:       []string{c.SubjectOrganization},
+			OrganizationalUnit: []string{c.SubjectOrganizationalUnit},
+			Locality:           []string{c.SubjectLocality},
+			CommonName:         c.SubjectCommonName,
+		},
+		NotBefore:          notAfter.Add(-time.Hour * 24),
```

Comment:
Document somewhere that all certificates have a validity period of 24 hours. Maybe make that a package var too? Do we expect all logs to accept certificates that aren't valid yet (but fall within their temporal interval, if they have one)?

Reply:
Good question, no idea whether we expect all Logs to accept certs that aren't valid yet.  My gut feeling is that we do, as they're certificates that could still be used in the future (unlike expired certificates that will never be usable again), although I'm not entirely sure...  I guess we'll find out pretty quickly whether Logs do actually accept them or not once we start submitting these to them :P